consensus/src/quorum_store/batch_generator.rs

```diff
@@ -99,112 +96,79 @@ impl BatchGenerator {
         }
     }
 
-    pub(crate) async fn handle_scheduled_pull(&mut self, max_count: u64) -> Option<Fragment> {
+    pub(crate) async fn handle_scheduled_pull(&mut self, max_count: u64) -> Option<Batch> {
         // TODO: as an optimization, we could filter out the txns that have expired
 
-        let mut exclude_txns: Vec<_> = self
+        let exclude_txns: Vec<_> = self
             .batches_in_progress
             .values()
             .flatten()
             .cloned()
             .collect();
-        exclude_txns.extend(self.batch_builder.summaries().clone());
 
         trace!("QS: excluding txs len: {:?}", exclude_txns.len());
-        let mut end_batch = false;
         let pulled_txns = self
             .mempool_proxy
             .pull_internal(
                 max_count,
                 self.config.mempool_txn_pull_max_bytes,
-                // allow creating non-full fragments
-                // is this a good place to disable fragments actually?
                 true,
                 exclude_txns,
             )
             .await
             .unwrap_or_default();
 
         trace!("QS: pulled_txns len: {:?}", pulled_txns.len());
+
         if pulled_txns.is_empty() {
             counters::PULLED_EMPTY_TXNS_COUNT.inc();
+            // Quorum store metrics
+            counters::CREATED_EMPTY_BATCHES_COUNT.inc();
+
+            let duration = self.last_end_batch_time.elapsed().as_secs_f64();
+            counters::EMPTY_BATCH_CREATION_DURATION
+                .observe_duration(Duration::from_secs_f64(duration));
+
+            self.last_end_batch_time = Instant::now();
+            return None;
         } else {
             counters::PULLED_TXNS_COUNT.inc();
             counters::PULLED_TXNS_NUM.observe(pulled_txns.len() as f64);
         }
 
-        if !self
-            .batch_builder
-            .append_transactions(&pulled_txns, max_count as usize)
```

Comment:
I don't understand the logic fully here, someone from QS should take a look, but this was checking if batch is full - to create it early.

so "true" on old line 122 (new line 115) should be changed to be equivalent to the second part of the condition here:
"self.last_end_batch_time.elapsed().as_millis() > self.config.end_batch_ms as u128"

but yeah, I am not sure how this ever slowed anything down, when end_batch_ms is 10, and poll outside is 25, hm?

@bchocho